### ISSUE TITLE
The implementation of pattern does not follow the relationship isomorphism

### ISSUE THREAD

**Issue by czpmango** [neo4j/neo4j#13018]
**Neo4j Version:** 4.2.4
**Operating System:** MacOs
**API:** Cypher

According to the documentation, the same path cannot contain the same relationship. Does this mean that variables of type relationship collection cannot appear in the same path matching?
```
Relationship isomorphism
The same relationship cannot be returned more than once for each path matching record. Cypher makes use of relationship isomorphism for path matching.
Neo4j Cypher makes use of relationship isomorphism for path matching and is a very effective way of reducing the result set size and preventing infinite traversals.
```

So I tested this statement and it behaves exactly as documented:
```
neo4j@neo4j> match (v1)-[e]->(v2)-[e]-(v3) return p; 
Cannot use the same relationship variable 'e' for multiple patterns (line 2, column 13 (offset: 14))
"match (v1)-[e]->(v2)-[e]-(v3) return p;"
```
But for the statement with the same semantic, or so I thought, I got different returns:
```
neo4j@neo4j> match p=(v1)-[e*1]->(v2)-[e*1]-(v3) return p; 
+-------------------------------------------+
| p                                         |
+-------------------------------------------+
| ({id: 2})-[:E]->({id: 1})<-[:E]-({id: 2}) |
| ({id: 3})-[:E]->({id: 1})<-[:E]-({id: 3}) |
| ({id: 4})-[:E]->({id: 1})<-[:E]-({id: 4}) |
| ({id: 1})-[:E]->({id: 2})<-[:E]-({id: 1}) |
| ({id: 1})-[:E]->({id: 3})<-[:E]-({id: 1}) |
| ({id: 1})-[:E]->({id: 4})<-[:E]-({id: 1}) |
| ({id: 4})-[:E]->({id: 4})-[:E]->({id: 4}) |
+-------------------------------------------+
```
So my **question 1** is, how does it make sense that the two statements above get different results? If the comparison of the relationships ignores the direction, shouldn't statement `match (v1)-[e]->(v2)-[e]-(v3) return p` also report an error?

I also tested this statement:
```
neo4j@neo4j> match p=(v1)-[e*1]->(v2)-[e*2]-(v3) return p,length(p); 
+---------------------------------------------------------------------------------------+
| p                                                                         | length(p) |
+---------------------------------------------------------------------------------------+
| ({id: 1})-[:E]->({id: 2})-[:E]->({id: 1})-[:E]->({id: 2})-[:E]->({id: 1}) | 4         |
| ({id: 1})-[:E]->({id: 3})-[:E]->({id: 1})-[:E]->({id: 3})-[:E]->({id: 1}) | 4         |
| ({id: 4})-[:E]->({id: 4})<-[:E]-({id: 4})-[:E]->({id: 4})<-[:E]-({id: 1}) | 4         |
| ({id: 1})-[:E]->({id: 4})-[:E]->({id: 1})-[:E]->({id: 4})-[:E]->({id: 1}) | 4         |
| ({id: 2})-[:E]->({id: 1})-[:E]->({id: 2})-[:E]->({id: 1})-[:E]->({id: 2}) | 4         |
| ({id: 3})-[:E]->({id: 1})-[:E]->({id: 3})-[:E]->({id: 1})-[:E]->({id: 3}) | 4         |
| ({id: 4})-[:E]->({id: 1})-[:E]->({id: 4})-[:E]->({id: 1})-[:E]->({id: 4}) | 4         |
+---------------------------------------------------------------------------------------+
```
As you can see, the same relationship returned more than once for each path matching record and the length of path is not as expected.
So **question 2**, is the result of this statement unexpected?

Similarly, such problems exist in directed edge patterns:
```
neo4j@neo4j> match p=(v1)-[e]->(v2)-[e]->(v3) return p,length(p); 
Cannot use the same relationship variable 'e' for multiple patterns (line 2, column 15 (offset: 16))
"match p=(v1)-[e]->(v2)-[e]->(v3) return p,length(p);"
               ^
neo4j@neo4j> match p=(v1)-[e*1]->(v2)-[e*1]->(v3) return p,length(p); 
+-------------------------------------------------------+
| p                                         | length(p) |
+-------------------------------------------------------+
| ({id: 4})-[:E]->({id: 4})-[:E]->({id: 4}) | 2         |
+-------------------------------------------------------+

1 row available after 24 ms, consumed after another 0 ms

neo4j@neo4j> match p=(v1)-[e*1]->(v2)-[e*2]->(v3) return p,length(p); 
+---------------------------------------------------------------------------------------+
| p                                                                         | length(p) |
+---------------------------------------------------------------------------------------+
| ({id: 1})-[:E]->({id: 2})-[:E]->({id: 1})-[:E]->({id: 2})-[:E]->({id: 1}) | 4         |
| ({id: 1})-[:E]->({id: 3})-[:E]->({id: 1})-[:E]->({id: 3})-[:E]->({id: 1}) | 4         |
| ({id: 1})-[:E]->({id: 4})-[:E]->({id: 1})-[:E]->({id: 4})-[:E]->({id: 1}) | 4         |
| ({id: 2})-[:E]->({id: 1})-[:E]->({id: 2})-[:E]->({id: 1})-[:E]->({id: 2}) | 4         |
| ({id: 3})-[:E]->({id: 1})-[:E]->({id: 3})-[:E]->({id: 1})-[:E]->({id: 3}) | 4         |
| ({id: 4})-[:E]->({id: 1})-[:E]->({id: 4})-[:E]->({id: 1})-[:E]->({id: 4}) | 4         |
+---------------------------------------------------------------------------------------+
```
**question 3**, do these statements behave as expected? What are the semantics of these statements?

---

I also tested some other statements:
```
neo4j@neo4j> match p=(v1)-[e*1..2]->(v2)-[e*2..4]->(v3) return length(p),p; 
+-------------------------------------------------------------------------------------------------------------------------------------------------------+
| length(p) | p                                                                                                                                         |
+-------------------------------------------------------------------------------------------------------------------------------------------------------+
| 4         | ({id: 1})-[:E]->({id: 2})-[:E]->({id: 1})-[:E]->({id: 2})-[:E]->({id: 1})                                                                 |
| 8         | ({id: 1})-[:E]->({id: 3})-[:E]->({id: 1})-[:E]->({id: 2})-[:E]->({id: 1})-[:E]->({id: 3})-[:E]->({id: 1})-[:E]->({id: 2})-[:E]->({id: 1}) |
| 8         | ({id: 1})-[:E]->({id: 4})-[:E]->({id: 1})-[:E]->({id: 2})-[:E]->({id: 1})-[:E]->({id: 4})-[:E]->({id: 1})-[:E]->({id: 2})-[:E]->({id: 1}) |
| 4         | ({id: 1})-[:E]->({id: 3})-[:E]->({id: 1})-[:E]->({id: 3})-[:E]->({id: 1})                                                                 |
| 8         | ({id: 1})-[:E]->({id: 2})-[:E]->({id: 1})-[:E]->({id: 3})-[:E]->({id: 1})-[:E]->({id: 2})-[:E]->({id: 1})-[:E]->({id: 3})-[:E]->({id: 1}) |
| 8         | ({id: 1})-[:E]->({id: 4})-[:E]->({id: 1})-[:E]->({id: 3})-[:E]->({id: 1})-[:E]->({id: 4})-[:E]->({id: 1})-[:E]->({id: 3})-[:E]->({id: 1}) |
| 4         | ({id: 1})-[:E]->({id: 4})-[:E]->({id: 1})-[:E]->({id: 4})-[:E]->({id: 1})                                                                 |
| 8         | ({id: 1})-[:E]->({id: 2})-[:E]->({id: 1})-[:E]->({id: 4})-[:E]->({id: 1})-[:E]->({id: 2})-[:E]->({id: 1})-[:E]->({id: 4})-[:E]->({id: 1}) |
| 8         | ({id: 1})-[:E]->({id: 3})-[:E]->({id: 1})-[:E]->({id: 4})-[:E]->({id: 1})-[:E]->({id: 3})-[:E]->({id: 1})-[:E]->({id: 4})-[:E]->({id: 1}) |
| 6         | ({id: 1})-[:E]->({id: 4})-[:E]->({id: 4})-[:E]->({id: 1})-[:E]->({id: 4})-[:E]->({id: 4})-[:E]->({id: 1})                                 |
| 4         | ({id: 2})-[:E]->({id: 1})-[:E]->({id: 2})-[:E]->({id: 1})-[:E]->({id: 2})                                                                 |
| 8         | ({id: 2})-[:E]->({id: 1})-[:E]->({id: 3})-[:E]->({id: 1})-[:E]->({id: 2})-[:E]->({id: 1})-[:E]->({id: 3})-[:E]->({id: 1})-[:E]->({id: 2}) |
| 8         | ({id: 2})-[:E]->({id: 1})-[:E]->({id: 4})-[:E]->({id: 1})-[:E]->({id: 2})-[:E]->({id: 1})-[:E]->({id: 4})-[:E]->({id: 1})-[:E]->({id: 2}) |
| 8         | ({id: 3})-[:E]->({id: 1})-[:E]->({id: 2})-[:E]->({id: 1})-[:E]->({id: 3})-[:E]->({id: 1})-[:E]->({id: 2})-[:E]->({id: 1})-[:E]->({id: 3}) |
| 4         | ({id: 3})-[:E]->({id: 1})-[:E]->({id: 3})-[:E]->({id: 1})-[:E]->({id: 3})                                                                 |
| 8         | ({id: 3})-[:E]->({id: 1})-[:E]->({id: 4})-[:E]->({id: 1})-[:E]->({id: 3})-[:E]->({id: 1})-[:E]->({id: 4})-[:E]->({id: 1})-[:E]->({id: 3}) |
| 8         | ({id: 4})-[:E]->({id: 1})-[:E]->({id: 2})-[:E]->({id: 1})-[:E]->({id: 4})-[:E]->({id: 1})-[:E]->({id: 2})-[:E]->({id: 1})-[:E]->({id: 4}) |
| 8         | ({id: 4})-[:E]->({id: 1})-[:E]->({id: 3})-[:E]->({id: 1})-[:E]->({id: 4})-[:E]->({id: 1})-[:E]->({id: 3})-[:E]->({id: 1})-[:E]->({id: 4}) |
| 4         | ({id: 4})-[:E]->({id: 1})-[:E]->({id: 4})-[:E]->({id: 1})-[:E]->({id: 4})                                                                 |
| 6         | ({id: 4})-[:E]->({id: 4})-[:E]->({id: 1})-[:E]->({id: 4})-[:E]->({id: 4})-[:E]->({id: 1})-[:E]->({id: 4})                                 |
| 6         | ({id: 4})-[:E]->({id: 1})-[:E]->({id: 4})-[:E]->({id: 4})-[:E]->({id: 1})-[:E]->({id: 4})-[:E]->({id: 4})                                 |
+-------------------------------------------------------------------------------------------------------------------------------------------------------+

21 rows available after 31 ms, consumed after another 3 ms
neo4j@neo4j> match p1=(v1)-[e*1..2]->(v2) match p2=(v2)-[e*2..4]->(v3) return length(p1)+length(p2),p1,p2; 
+-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| length(p1)+length(p2) | p1                                                                        | p2                                                                        |
+-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| 4                     | ({id: 1})-[:E]->({id: 2})-[:E]->({id: 1})                                 | ({id: 1})-[:E]->({id: 2})-[:E]->({id: 1})                                 |
| 8                     | ({id: 1})-[:E]->({id: 3})-[:E]->({id: 1})-[:E]->({id: 2})-[:E]->({id: 1}) | ({id: 1})-[:E]->({id: 3})-[:E]->({id: 1})-[:E]->({id: 2})-[:E]->({id: 1}) |
| 8                     | ({id: 1})-[:E]->({id: 4})-[:E]->({id: 1})-[:E]->({id: 2})-[:E]->({id: 1}) | ({id: 1})-[:E]->({id: 4})-[:E]->({id: 1})-[:E]->({id: 2})-[:E]->({id: 1}) |
| 4                     | ({id: 1})-[:E]->({id: 3})-[:E]->({id: 1})                                 | ({id: 1})-[:E]->({id: 3})-[:E]->({id: 1})                                 |
| 8                     | ({id: 1})-[:E]->({id: 2})-[:E]->({id: 1})-[:E]->({id: 3})-[:E]->({id: 1}) | ({id: 1})-[:E]->({id: 2})-[:E]->({id: 1})-[:E]->({id: 3})-[:E]->({id: 1}) |
| 8                     | ({id: 1})-[:E]->({id: 4})-[:E]->({id: 1})-[:E]->({id: 3})-[:E]->({id: 1}) | ({id: 1})-[:E]->({id: 4})-[:E]->({id: 1})-[:E]->({id: 3})-[:E]->({id: 1}) |
| 4                     | ({id: 1})-[:E]->({id: 4})-[:E]->({id: 1})                                 | ({id: 1})-[:E]->({id: 4})-[:E]->({id: 1})                                 |
| 8                     | ({id: 1})-[:E]->({id: 2})-[:E]->({id: 1})-[:E]->({id: 4})-[:E]->({id: 1}) | ({id: 1})-[:E]->({id: 2})-[:E]->({id: 1})-[:E]->({id: 4})-[:E]->({id: 1}) |
| 8                     | ({id: 1})-[:E]->({id: 3})-[:E]->({id: 1})-[:E]->({id: 4})-[:E]->({id: 1}) | ({id: 1})-[:E]->({id: 3})-[:E]->({id: 1})-[:E]->({id: 4})-[:E]->({id: 1}) |
| 6                     | ({id: 1})-[:E]->({id: 4})-[:E]->({id: 4})-[:E]->({id: 1})                 | ({id: 1})-[:E]->({id: 4})-[:E]->({id: 4})-[:E]->({id: 1})                 |
| 4                     | ({id: 2})-[:E]->({id: 1})-[:E]->({id: 2})                                 | ({id: 2})-[:E]->({id: 1})-[:E]->({id: 2})                                 |
| 8                     | ({id: 2})-[:E]->({id: 1})-[:E]->({id: 3})-[:E]->({id: 1})-[:E]->({id: 2}) | ({id: 2})-[:E]->({id: 1})-[:E]->({id: 3})-[:E]->({id: 1})-[:E]->({id: 2}) |
| 8                     | ({id: 2})-[:E]->({id: 1})-[:E]->({id: 4})-[:E]->({id: 1})-[:E]->({id: 2}) | ({id: 2})-[:E]->({id: 1})-[:E]->({id: 4})-[:E]->({id: 1})-[:E]->({id: 2}) |
| 8                     | ({id: 3})-[:E]->({id: 1})-[:E]->({id: 2})-[:E]->({id: 1})-[:E]->({id: 3}) | ({id: 3})-[:E]->({id: 1})-[:E]->({id: 2})-[:E]->({id: 1})-[:E]->({id: 3}) |
| 4                     | ({id: 3})-[:E]->({id: 1})-[:E]->({id: 3})                                 | ({id: 3})-[:E]->({id: 1})-[:E]->({id: 3})                                 |
| 8                     | ({id: 3})-[:E]->({id: 1})-[:E]->({id: 4})-[:E]->({id: 1})-[:E]->({id: 3}) | ({id: 3})-[:E]->({id: 1})-[:E]->({id: 4})-[:E]->({id: 1})-[:E]->({id: 3}) |
| 8                     | ({id: 4})-[:E]->({id: 1})-[:E]->({id: 2})-[:E]->({id: 1})-[:E]->({id: 4}) | ({id: 4})-[:E]->({id: 1})-[:E]->({id: 2})-[:E]->({id: 1})-[:E]->({id: 4}) |
| 8                     | ({id: 4})-[:E]->({id: 1})-[:E]->({id: 3})-[:E]->({id: 1})-[:E]->({id: 4}) | ({id: 4})-[:E]->({id: 1})-[:E]->({id: 3})-[:E]->({id: 1})-[:E]->({id: 4}) |
| 4                     | ({id: 4})-[:E]->({id: 1})-[:E]->({id: 4})                                 | ({id: 4})-[:E]->({id: 1})-[:E]->({id: 4})                                 |
| 6                     | ({id: 4})-[:E]->({id: 4})-[:E]->({id: 1})-[:E]->({id: 4})                 | ({id: 4})-[:E]->({id: 4})-[:E]->({id: 1})-[:E]->({id: 4})                 |
| 6                     | ({id: 4})-[:E]->({id: 1})-[:E]->({id: 4})-[:E]->({id: 4})                 | ({id: 4})-[:E]->({id: 1})-[:E]->({id: 4})-[:E]->({id: 4})                 |
+-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+

21 rows available after 2 ms, consumed after another 7 ms

```
My **question 4** is that are the above two statements semantically equivalent? 
IMHO, the semantics of the last statement is clear and follows `relationship isomorphism` exactly.

**Comment by HannesSandberg** [comment]
Thank you for your report, we will investigate this!

**Comment by sherfert** [comment]
Hi @czpmango 

You are right and this is a bug. The bug is already fixed in neo4j 5.1 and newer versions of the neo4j 5 series.
We will also fix it in neo4j 4.4. We will not backport it to neo4j 4.2, that version is no longer supported.

Another thing to consider is that we are going to disallow repeating a variable length relationship variable in a future release, since it has unclear semantics in many situations.

Regarding your questions:
1. `match p=(v1)-[e*1]->(v2)-[e*1]-(v3) return p` returns no results in `5.3`. It does not return an error, mostly for backwards-compatibility reasons. As mentioned above, repeating `e` will be forbidden in a future release
2. `match p=(v1)-[e*1]->(v2)-[e*2]-(v3) return p,length(p)` also returns no results in `5.3`.
3. Same.
4. In your queries, the 2 occurences of `e` use different quantifiers, i.e. `[e*1..2]` vs `[e*2..4]`. This is one of the situations where the semantics are unclear and the product does not not behave consistently. As you can see, in the last results, `p1` also contains paths of length 4, even though the quantifier in `p1` is `1..2`. This will be solved by disallow repeating a variable length relationship variable in a future release, as mentioned above. Until then, a warning is present in `5.3`.


I will close this issue when the fix to 4.4 has been merged.

**Comment by czpmango** [comment]
@sherfert Thank you for such a detailed reply!

I test the new 5.3 version, as you said, the above issues about single path matching have been fixed. 

But I don‘t get the point about what you said `“This is one of the situations where the semantics are unclear and the product does not not behave consistently”`. Why disallow repeating a variable length relationship variable in the future version? IMHO, the semantics are clear.

For example, how to implement the following statement, or is there another equivalent syntax to describe this scene?
```
match p1=({id:4})--(v1)-[e*1..3]->(v2) optional match p2=(v2:A)-[e*2..4]-(v3) return p1,p2;
```

IMHO, repeated variables of variable-length relationships are still a useful feature.
`match p1=({id:4})--(v1)-[e*1..3]->(v2) match p2=(v1:A)-[e*2..4]-(v3) return p1, p2` and `match p1=({id:4})--(v1:A)-[e*2..3]->(v2) return p1` are approximately semantically equivalent(Am I right? I'm not really sure). But what cypher syntax should I use to describe the scene containing optional match without this feature?


As far as I know, the gql standard also prohibits implicit equi-joins between group variables, which like variable-length relationships.
Am I missing something, or do you have other considerations?

**Comment by czpmango** [comment]
FYI, some related bugs in 5.3 version:
```
neo4j@neo4j> match p1=(v1)-[e*1..3]->(v2) match p2=(v2:A)-[e*2..4]-(v3) where length(p2)<2 RETURN p2;
+------------------------------+
| p2                           |
+------------------------------+
| (:A {id: 2})<-[:E]-({id: 1}) |
| (:A {id: 2})<-[:E]-({id: 4}) |
+------------------------------+

2 rows
ready to start consuming query after 1 ms, results consumed after another 2 ms
neo4j@neo4j> match p1=({id:4})--(v1)-[e*1..3]->(v2) optional match p2=(v2:A)-[e*2..4]-(v3) return p1,p2;
+----------------------------------------------------------------------------------------------------------------+
| p1                                                              | p2                                           |
+----------------------------------------------------------------------------------------------------------------+
| ({id: 4})<-[:E]-({id: 3})-[:E]->({id: 1})                       | NULL                                         |
| ({id: 4})<-[:E]-({id: 3})<-[:E]-(:A {id: 2})-[:E]->(:A {id: 2}) | (:A {id: 2})<-[:E]-({id: 1})<-[:E]-({id: 3}) |
Invalid usage of PathValueBuilder, 3 must be a node in -[0]-
```

**Comment by sherfert** [comment]
> But I don‘t get the point about what you said `“This is one of the situations where the semantics are unclear and the product does not not behave consistently”`. Why disallow repeating a variable length relationship variable in the future version? IMHO, the semantics are clear.

Our documentation does not explain the semantics anywhere and I think we have different interpretations.
If you revisit the following query you posted:

```
21 rows available after 31 ms, consumed after another 3 ms
neo4j@neo4j> match p1=(v1)-[e*1..2]->(v2) match p2=(v2)-[e*2..4]->(v3) return length(p1)+length(p2),p1,p2; 
+-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| length(p1)+length(p2) | p1                                                                        | p2                                                                        |
+-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| 4                     | ({id: 1})-[:E]->({id: 2})-[:E]->({id: 1})                                 | ({id: 1})-[:E]->({id: 2})-[:E]->({id: 1})                                 |
| 8                     | ({id: 1})-[:E]->({id: 3})-[:E]->({id: 1})-[:E]->({id: 2})-[:E]->({id: 1}) | ({id: 1})-[:E]->({id: 3})-[:E]->({id: 1})-[:E]->({id: 2})-[:E]->({id: 1}) |
| 8                     | ({id: 1})-[:E]->({id: 4})-[:E]->({id: 1})-[:E]->({id: 2})-[:E]->({id: 1}) | ({id: 1})-[:E]->({id: 4})-[:E]->({id: 1})-[:E]->({id: 2})-[:E]->({id: 1}) |
| 4                     | ({id: 1})-[:E]->({id: 3})-[:E]->({id: 1})                                 | ({id: 1})-[:E]->({id: 3})-[:E]->({id: 1})                                 |
| 8                     | ({id: 1})-[:E]->({id: 2})-[:E]->({id: 1})-[:E]->({id: 3})-[:E]->({id: 1}) | ({id: 1})-[:E]->({id: 2})-[:E]->({id: 1})-[:E]->({id: 3})-[:E]->({id: 1}) |
| 8                     | ({id: 1})-[:E]->({id: 4})-[:E]->({id: 1})-[:E]->({id: 3})-[:E]->({id: 1}) | ({id: 1})-[:E]->({id: 4})-[:E]->({id: 1})-[:E]->({id: 3})-[:E]->({id: 1}) |
| 4                     | ({id: 1})-[:E]->({id: 4})-[:E]->({id: 1})                                 | ({id: 1})-[:E]->({id: 4})-[:E]->({id: 1})                                 |
| 8                     | ({id: 1})-[:E]->({id: 2})-[:E]->({id: 1})-[:E]->({id: 4})-[:E]->({id: 1}) | ({id: 1})-[:E]->({id: 2})-[:E]->({id: 1})-[:E]->({id: 4})-[:E]->({id: 1}) |
| 8                     | ({id: 1})-[:E]->({id: 3})-[:E]->({id: 1})-[:E]->({id: 4})-[:E]->({id: 1}) | ({id: 1})-[:E]->({id: 3})-[:E]->({id: 1})-[:E]->({id: 4})-[:E]->({id: 1}) |
| 6                     | ({id: 1})-[:E]->({id: 4})-[:E]->({id: 4})-[:E]->({id: 1})                 | ({id: 1})-[:E]->({id: 4})-[:E]->({id: 4})-[:E]->({id: 1})                 |
| 4                     | ({id: 2})-[:E]->({id: 1})-[:E]->({id: 2})                                 | ({id: 2})-[:E]->({id: 1})-[:E]->({id: 2})                                 |
| 8                     | ({id: 2})-[:E]->({id: 1})-[:E]->({id: 3})-[:E]->({id: 1})-[:E]->({id: 2}) | ({id: 2})-[:E]->({id: 1})-[:E]->({id: 3})-[:E]->({id: 1})-[:E]->({id: 2}) |
| 8                     | ({id: 2})-[:E]->({id: 1})-[:E]->({id: 4})-[:E]->({id: 1})-[:E]->({id: 2}) | ({id: 2})-[:E]->({id: 1})-[:E]->({id: 4})-[:E]->({id: 1})-[:E]->({id: 2}) |
| 8                     | ({id: 3})-[:E]->({id: 1})-[:E]->({id: 2})-[:E]->({id: 1})-[:E]->({id: 3}) | ({id: 3})-[:E]->({id: 1})-[:E]->({id: 2})-[:E]->({id: 1})-[:E]->({id: 3}) |
| 4                     | ({id: 3})-[:E]->({id: 1})-[:E]->({id: 3})                                 | ({id: 3})-[:E]->({id: 1})-[:E]->({id: 3})                                 |
| 8                     | ({id: 3})-[:E]->({id: 1})-[:E]->({id: 4})-[:E]->({id: 1})-[:E]->({id: 3}) | ({id: 3})-[:E]->({id: 1})-[:E]->({id: 4})-[:E]->({id: 1})-[:E]->({id: 3}) |
| 8                     | ({id: 4})-[:E]->({id: 1})-[:E]->({id: 2})-[:E]->({id: 1})-[:E]->({id: 4}) | ({id: 4})-[:E]->({id: 1})-[:E]->({id: 2})-[:E]->({id: 1})-[:E]->({id: 4}) |
| 8                     | ({id: 4})-[:E]->({id: 1})-[:E]->({id: 3})-[:E]->({id: 1})-[:E]->({id: 4}) | ({id: 4})-[:E]->({id: 1})-[:E]->({id: 3})-[:E]->({id: 1})-[:E]->({id: 4}) |
| 4                     | ({id: 4})-[:E]->({id: 1})-[:E]->({id: 4})                                 | ({id: 4})-[:E]->({id: 1})-[:E]->({id: 4})                                 |
| 6                     | ({id: 4})-[:E]->({id: 4})-[:E]->({id: 1})-[:E]->({id: 4})                 | ({id: 4})-[:E]->({id: 4})-[:E]->({id: 1})-[:E]->({id: 4})                 |
| 6                     | ({id: 4})-[:E]->({id: 1})-[:E]->({id: 4})-[:E]->({id: 4})                 | ({id: 4})-[:E]->({id: 1})-[:E]->({id: 4})-[:E]->({id: 4})                 |
+-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
```

Here the variable `e` appears in two places, but with different quantifiers. First with `1..2` and then with `2..4`. But `e` is only a single variable, so both quantifiers would need to be applied, effectively giving `2..2`, i.e. only length 2. As you can see the product currently also finds paths of length 3 and 4. I consider this to be wrong.

I am not sure how you would define your semantics, but they seem to be different from the semantics we would want to achieve here.

---

> For example, how to implement the following statement, or is there another equivalent syntax to describe this scene?
> 
> ```
> match p1=({id:4})--(v1)-[e*1..3]->(v2) optional match p2=(v2:A)-[e*2..4]-(v3) return p1,p2;
> ```

In this example, `e` is again repeated with 2 different quantifiers. I think you would want to rewrite this, e.g. to
```
match p1=({id:4})--(v1)-[e*1..3]->(v2) optional match p2=(v2:A)--()-[e*1..3]-(v3) return p1,p2;
```
?


---


> IMHO, repeated variables of variable-length relationships are still a useful feature. `match p1=({id:4})--(v1)-[e*1..3]->(v2) match p2=(v1:A)-[e*2..4]-(v3) return p1, p2` and `match p1=({id:4})--(v1:A)-[e*2..3]->(v2) return p1` are approximately semantically equivalent(Am I right? I'm not really sure). But what cypher syntax should I use to describe the scene containing optional match without this feature?

I agree, reusing a repeated var-length relationship is useful together with path variables, because it essentially, allows to reuse a segment of a path in 2 path variables, e.g.

```
MATCH p1 = (a:A)--()-[e*3..5]->()--(b:B)
MATCH p2 = (c:C)--()-[e*3..5]->()--(d:D)
RETURN p1, p2
```

We have currently only deprecated this. Maybe I was a bit hasty to state we will definitely remove this in the future.
If we remove this, we will provide alternative syntax for this use case.
 
---


> As far as I know, the gql standard also prohibits implicit equi-joins between group variables, which like variable-length relationships. Am I missing something, or do you have other considerations?

I cannot answer questions on the GQL standard here, unfortunately.

**Comment by sherfert** [comment]
> FYI, some related bugs in 5.3 version:
> 
> ```
> neo4j@neo4j> match p1=(v1)-[e*1..3]->(v2) match p2=(v2:A)-[e*2..4]-(v3) where length(p2)<2 RETURN p2;
> +------------------------------+
> | p2                           |
> +------------------------------+
> | (:A {id: 2})<-[:E]-({id: 1}) |
> | (:A {id: 2})<-[:E]-({id: 4}) |
> +------------------------------+
> 
> 2 rows
> ready to start consuming query after 1 ms, results consumed after another 2 ms
> neo4j@neo4j> match p1=({id:4})--(v1)-[e*1..3]->(v2) optional match p2=(v2:A)-[e*2..4]-(v3) return p1,p2;
> +----------------------------------------------------------------------------------------------------------------+
> | p1                                                              | p2                                           |
> +----------------------------------------------------------------------------------------------------------------+
> | ({id: 4})<-[:E]-({id: 3})-[:E]->({id: 1})                       | NULL                                         |
> | ({id: 4})<-[:E]-({id: 3})<-[:E]-(:A {id: 2})-[:E]->(:A {id: 2}) | (:A {id: 2})<-[:E]-({id: 1})<-[:E]-({id: 3}) |
> Invalid usage of PathValueBuilder, 3 must be a node in -[0]-
> ```

A fix for the original issue was now merged into 4.4 and will appear in the next 4.4 release.

@czpmango Could you create separate issues for the bugs you found in 5.3? Please also provide a small setup, i.e. some `CREATE` statements, to reproduce the behavior. Thank you!

**Comment by czpmango** [comment]
> > For example, how to implement the following statement, or is there another equivalent syntax to describe this scene?
match p1=({id:4})--(v1)-[e*1..3]->(v2) optional match p2=(v2:A)-[e*2..4]-(v3) return p1,p2;

> In this example, e is again repeated with 2 different quantifiers. I think you would want to rewrite this, e.g. to
> > match p1=({id:4})--(v1)-[e*1..3]->(v2) optional match p2=(v2:A)--()-[e*1..3]-(v3) return p1,p2;

No, you misunderstood the semantics of this query. The variables `v2` and `e` are common variables of these two patterns and they are linked. The label expression `v2:A` appears in the optional match pattern but will not affect the cardinality of the match query result, and the value of `v2` in `p1` that does not satisfy `v2:A` will be null, the same is true for `e`.  The semantics of this query may be closer to `match p1=({id:4})--(v1)-[e*1..3]->(v2) optional match p2=(v2:A)<-[e*2..3]-(v3) return p1,p2` but it is indeed different from `match p1=({id:4})--(v1)-[e*1..3]->(v2) optional match p2=(v2:A)--()-[e*1..3]-(v3) return p1,p2`.

> I agree, reusing a repeated var-length relationship is useful together with path variables, because it essentially, allows to reuse a segment of a path in 2 path variables, e.g.
> We have currently only deprecated this. Maybe I was a bit hasty to state we will definitely remove this in the future.
If we remove this, we will provide alternative syntax for this use case.

Basically ACK and look forward to your release of new features to support this scene.

**Comment by czpmango** [comment]
> Could you create separate issues for the bugs you found in 5.3? Please also provide a small setup, i.e. some `CREATE` statements, to reproduce the behavior. Thank you!

https://github.com/neo4j/neo4j/issues/13029
